File: domain/src/main/java/com/thinkzi/weather/domain/usecase/base/AsynchronousRxUseCase.java
package com.thinkzi.weather.domain.usecase.base;

import com.thinkzi.weather.domain.executor.PostExecutionThread;
import com.thinkzi.weather.domain.executor.ThreadExecutor;
import com.thinkzi.weather.domain.utility.check.Checker;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

/**
 * provide common abstract execution of all reactive asynchronous UseCase(Clean Architecture)
 * */
public abstract class AsynchronousRxUseCase {

    // executor thread pool to execute work
    final ThreadExecutor _threadExecutor;

    // thread created to change the execution context
    final PostExecutionThread _postExecutionThread;

    // observer set for work result
    private final CompositeDisposable _compositeDisposables;

    AsynchronousRxUseCase(ThreadExecutor _threadExecutor, PostExecutionThread _postExecutionThread) {

        this._threadExecutor = _threadExecutor;
        this._postExecutionThread = _postExecutionThread;
        this._compositeDisposables = new CompositeDisposable();

    }

    /**
     * dispose all observers
     * */
    public void dispose() {

        if (!_compositeDisposables.isDisposed()) {

            _compositeDisposables.dispose();

        }

    }

    /**
     * add a observer
     * */
    void addDisposable(Disposable _disposable) {

        Checker.checkNotNull(_disposable);
        Checker.checkNotNull(_compositeDisposables);

        _compositeDisposables.add(_disposable);

    }
}

File: domain/src/main/java/com/thinkzi/weather/domain/usecase/base/AsynchronousRxUseCase.kt
package com.thinkzi.weather.domain.usecase.base

import com.thinkzi.weather.domain.executor.PostExecutionThread
import com.thinkzi.weather.domain.executor.ThreadExecutor
import com.thinkzi.weather.domain.utility.check.Checker

import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

/**
 * provide common abstract execution of all reactive asynchronous UseCase(Clean Architecture)
 */
abstract class AsynchronousRxUseCase internal constructor(// executor thread pool to execute work
    internal val _threadExecutor: ThreadExecutor, // thread created to change the execution context
    internal val _postExecutionThread: PostExecutionThread
) {

    // observer set for work result
    private val _compositeDisposables: CompositeDisposable

    init {
        this._compositeDisposables = CompositeDisposable()

    }

    /**
     * dispose all observers
     */
    fun dispose() {

        if (!_compositeDisposables.isDisposed) {

            _compositeDisposables.dispose()

        }

    }

    /**
     * add a observer
     */
    internal fun addDisposable(_disposable: Disposable) {

        Checker.checkNotNull(_disposable)
        Checker.checkNotNull(_compositeDisposables)

        _compositeDisposables.add(_disposable)

    }
}
